src/linker/Linker/Driver.cs

```diff
@@ -789,14 +790,18 @@ public int Run (ILogger? customLogger = null, bool throwOnFatalLinkerException =
 				if (throwOnFatalLinkerException)
 					throw;
 				return lex.MessageContainer.Code ?? 1;
+			}
+			catch (ArgumentException) {
+				Context.LogError (null, DiagnosticId.LinkerUnexpectedError);
+				throw;
 			} catch (ResolutionException e) {
 				Context.LogError (null, DiagnosticId.FailedToResolveMetadataElement, e.Message);
-			} catch (Exception) {
+			} catch (Exception e) {
 				// Unhandled exceptions are usually linker bugs. Ask the user to report it.
 				Context.LogError (null, DiagnosticId.LinkerUnexpectedError);
+				Environment.FailFast ("Unexpected Error", e);
```

Comment:
Are we sure we want to take down the whole process?

Reply:
The goal of this change is to make it possible to diagnose hard to repro failures (like the ones we see in runtime CI currently). What this is trying to do is to crash the process in such a way that dump will be captured... and we can then investigate the dump.

We shortly discussed the possibility to use non fatal Watson dump on Windows, but for now decided that the hard crash would be better (it works everywhere, not just Windows).

Reply:
Right just ensure that all the products that integrate linker (VS, VS4M, VSCode, etc.) via msbuild are prepared for that.

Reply:
All of those should do it through msbuild - and in that case linker should run as an external process. If it crashes msbuild will produce an error - it doesn't look exactly pretty, but that's basically the tradeoff here - between pretty failures and diagnosable ones.

Reply:
Why do we need to catch these exceptions at all?

If we don't have a catch anywhere, it will just go into the runtime-default unhandled exception paths that also trigger Watson.

For example here we could do the equivalent of:

```csharp
try
{
    ...
}
catch when (PrintMessageToContactLinkerDevsAndReturnFalse()) { /* unreachable */ }
```

The difference is basically that if you have FailFast, the stack trace that the runtime prints to stderr is the one of the FailFast. If you don't catch the exception, the stack trace printed to stderr is the one for when the exception was thrown.

Reply:
Sorry for not marking this as still in progress, after discussing with Sven and Andy I realized we want to fail in the exception filter like Michal suggested to get the Watson dumps. I've updated the code to do that.

Reply:
My question was more about "why do we catch the exception in the first place when we don't handle it?". If we catch unexpected things, we need to add extra code to do the FailFast - but we could just be more mindful of what we catch instead.

Like in this specific spot I'm commenting on: we catch only to print something into the logger (and then we retrhow previously, or FailFast with this change). But if we instead print into the logger from a filter clause (a filter that returns `false` to not actually trigger a catch) we don't need to worry about rethrow or FailFast - the exceptions will fly past this spot.